tests/unit/fem/ptest_tdofs.cpp

```diff
@@ -0,0 +1,1012 @@
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "mfem.hpp"
+
+#include "catch.hpp"
+
+#include <fstream>
+#include <sstream>
+#include <iostream>
+
+/**
+ * \file ptest_tdofs
+ *
+ * Some tests related to parallel finite element spaces
+ *
+ * Specifically, computes global IDs for the vertices, edges,
+ * faces, elements and boundary elements of an mfem ParMesh.
+ *
+ * \note This test requires MPI and Hypre
+ */
+
+// define helper macros to print to cout from a given rank?
+#define PTEST_TDOF_USE_DEBUG_PRINT 0
+
+#ifdef MFEM_USE_MPI
+
+namespace
+{
+
+#if PTEST_TDOF_USE_DEBUG_PRINT
+#define MPRINT_FLUSH                          \
+   do {                                       \
+      MPI_Barrier(MPI_COMM_WORLD);            \
+      std::cout.flush();                      \
+       MPI_Barrier(MPI_COMM_WORLD);           \
+   } while(0)
+
+#define MPRINT(_rank,_msg)                    \
+    do {                                      \
+          std::cout << "[" << _rank << "]"    \
+                    << _msg                   \
+                    << std::endl;             \
+    } while(0)
+
+#define MPRINT_MASTER(_rank,_msg)             \
+    do {                                      \
+        MPRINT_FLUSH;                         \
+        if(_rank == 0) { MPRINT(_rank,_msg);} \
+        MPRINT_FLUSH;                         \
+    } while(0)
+
+#else  // PTEST_TDOF_USE_DEBUG_PRINT
+#define MPRINT_FLUSH                do{}while(0)
+#define MPRINT(_rank,_msg)          do{}while(0)
+#define MPRINT_MASTER(_rank,_msg)   do{}while(0)
+#endif //PTEST_TDOF_USE_DEBUG_PRINT
+
+// Define some parameters for the tests
+const int NUM_SER_REF_LEVELS = 1;
+const int NUM_PAR_REF_LEVELS = 0;
+const bool VERBOSE_OUTPUT = false;
+
+/// Helper function to refine the mesh uniformly
+inline void refineMesh(mfem::Mesh* mesh, int num_levels)
+{
+   for (int l = 0; l < num_levels; ++l)
+   {
+      mesh->UniformRefinement();
+   }
+}
+
+/// Helper function to collect some stats about the mesh
+inline std::string getMeshStats(mfem::Mesh* mesh)
+{
+   std::stringstream sstr;
+
+   sstr << "Mesh has: "
+        << "\n\t"<< mesh->GetNV() << " vertices."
+        << "\n\t"<< mesh->GetNEdges() << " edges."
+        << "\n\t"<< mesh->GetNFaces() << " faces."
+        << "\n\t"<< mesh->GetNE() << " elements."
+        << "\n\t"<< mesh->GetNBE() << " boundary elements.";
+
+   return sstr.str();
+}
+
+/// Helper to remove encoded sign from a DOF
+/// Adapted from protected static function in mfem::FiniteElementSpace
+inline int decodeDof(int dof)
```

Comment:
I don't entirely understand which FE spaces have encoded DOFs, but this was necessary here the RT spaces.

Any reason not to make this static function public in ``FiniteElementSpace`` (it is currently protected)?
See: https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/fespace.hpp#L143-L145

Reply:
In H(curl) and H(div), vector-valued shape functions may sometimes need to be multiplied by -1 to match each other across edges and faces. This is a transformation that cannot be expressed as DOF permutation (as in H1). Negative DOFs is a slightly hacky way of encoding this transformation, and should ideally be an internal thing, that's why I am hesitant to make this function public.